contributor-docs/code-change-guide.md

```diff
@@ -0,0 +1,516 @@
+Gradle project, tests can be executed via a Gradle command line, see below.
+
+## Console (shell) setup
+
+Equivalent command line:
+
+```shell
+$ cd beam
+$ ./gradlew :examples:java:wordCount
+```
+
+Upon finishing, one should see the following Gradle build log:
+
+```
+...
+BUILD SUCCESSFUL in 2m 32s
+96 actionable tasks: 9 executed, 87 up-to-date
+3:41:06 PM: Execution finished 'wordCount'.
+```
+
+and checking the output file:
+
+```shell
+
+$ head /tmp/output.txt*
+==> /tmp/output.txt-00000-of-00003 <==
+should: 38
+bites: 1
+depraved: 1
+gauntlet: 1
+battle: 6
+sith: 2
+cools: 1
+natures: 1
+hedge: 1
+words: 9
+
+==> /tmp/output.txt-00001-of-00003 <==
+elements: 1
+Advise: 2
+fearful: 2
+towards: 4
+ready: 8
+pared: 1
+left: 8
+safe: 4
+canst: 7
+warrant: 2
+
+==> /tmp/output.txt-00002-of-00003 <==
+chanced: 1
+...
+```
+
+*What does this command do?*
+
+It compiles the beam SDK and the word count pipeline (Hello-world program for
+data processing), then run the pipeline on Direct Runner.
+
+## Run a unit test
+
+Now, suppose you have made a code change in Java SDK (e.g. in `sdks/java/io/jdbc`),
+and want to run relevant unit tests locally to verify the change.
+
+Unit tests are under `src/test/java` folder of each project with filename `.../**Test.java` for unit tests and, `.../**IT.java` for integration tests. For example,
+
+* Run all unit tests under a project
+  ```
+  ./gradlew :sdks:java:harness:test
+  ```
+  Then JUnit report (in html web page) can be found under `<invoked_project>/build/reports/tests/test/index.html`
+
+* Run a specific test
+  ```
+  ./gradlew :sdks:java:harness:test --tests org.apache.beam.fn.harness.CachesTest
+  ./gradlew :sdks:java:harness:test --tests *CachesTest
+  ./gradlew :sdks:java:harness:test --tests *CachesTest.testClearableCache
+  ```
+
+* Run a specific test with IntelliJ
+  <img width="452" alt="image" src="https://github.com/apache/beam/assets/8010435/7ae2a65c-a104-48a2-8bad-ff8c52dd1943">
+  Just click the ticks to run a whole test class or a specific test. One can also set breakpoints and debug the test.
+
+* Exception: `sdks:java:core`. Many unit tests are invoked by task  `:runners:direct-java:needsRunnerTest` instead.
+
+  Reason: Java core does not depend on a runner. For unit test that executes a pipeline, it needs direct runner.
+
+* Run an integration test on direct runner
+
+## Run an integration test (*IT.java)
+
+An integration test uses `TestPipeline` where its option is set by `TestPipelineOptions` ([code path](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java))
+
+Differences from normal Pipeline:
+* Block on run (on TestDataflowRunner) by default
+* Has a default timeout (15 min)
+* Pipeline options set in System property `beamTestPipelineOptions`
+
+The last point is important: the test is configured by setting `-DbeamTestPipelineOptions=[...]`, including which runner it will be run.
```

Comment:
```suggestion
Note the final different, because you need to configure the test by setting `-DbeamTestPipelineOptions=[...]`. This property is where you set the runner to use.
```